src/kubernetes/multi_response_decoder.rs

```diff
@@ -0,0 +1,410 @@
+//! Decode multiple [`Response`]s.
+
+use k8s_openapi::{http::StatusCode, Response, ResponseError};
+
+/// Provides an algorithm to parse multiple [`Response`]s from multiple chunks
+/// of data represented as `&[u8]`.
+#[derive(Debug)]
+pub struct MultiResponseDecoder<T> {
+    pending_data: Vec<u8>,
+    responses_buffer: Vec<Result<T, ResponseError>>,
+}
+
+impl<T> MultiResponseDecoder<T>
+where
+    T: Response,
+{
+    /// Create a new [`MultiResponseDecoder`].
+    pub fn new() -> Self {
+        Self {
+            pending_data: Vec::new(),
+            responses_buffer: Vec::new(),
+        }
+    }
+
+    /// Take the next chunk of data and spit out parsed `T`s.
+    pub fn process_next_chunk(
+        &mut self,
+        chunk: &[u8],
+    ) -> std::vec::Drain<'_, Result<T, ResponseError>> {
+        self.pending_data.extend_from_slice(chunk);
+        loop {
+            match T::try_from_parts(StatusCode::OK, &self.pending_data) {
+                Ok((response, consumed_bytes)) => {
+                    debug_assert!(consumed_bytes > 0, "parser must've consumed some data");
+                    self.pending_data.drain(..consumed_bytes);
```

Comment:
This is designed for streaming processing, so we have to remove bytes that we consumed rather than letting them lying around.
I looked into it a bit deeper, and it looks like `Drain` doesn't deallocate the unused space. We should probably do it here though, at some point.
Also, we should probably consider adding using a ring buffer here.

There has to be a test to ensure this code properly frees memory, I'll add it.

Reply:
Added the test: https://github.com/timberio/vector/pull/2653/commits/72303228704d4a4d9b6a6a8c00ac6699cb5029bb
No signs of issues.

Reply:
Ok, great that you looked into `Drain`.